src/main/kotlin/com/atlassian/performance/tools/report/api/judge/Verdict.kt

```diff
@@ -14,15 +15,39 @@ class Verdict internal constructor(
     val reports: List<JUnitReport>,
     val failedActions: List<ActionType<*>>
 ) {
+    private val logger = LogManager.getLogger(this::class.java)
 
+    @Deprecated(message = "Use Builder instead", replaceWith = ReplaceWith("Verdict.Builder(reports).build()"))
     constructor(
         reports: List<JUnitReport>
     ) : this(
         reports = reports,
         failedActions = emptyList()
     )
 
-    private val logger = LogManager.getLogger(this::class.java)
+    /**
+     * Since 3.12.0
+     */
+    class Builder(reports: List<JUnitReport>) {
+        private var failedActions: MutableList<ActionType<*>> = ArrayList()
+        private var reports: MutableList<JUnitReport> = ArrayList(reports)
+
+        fun failedActions(failedActions: List<ActionType<*>>): Builder = apply {
+            this.failedActions = ArrayList(failedActions)
+        }
+
+        fun addReports(reports: List<JUnitReport>): Builder = apply {
+            this.reports.addAll(reports)
+        }
+
+        fun build(): Verdict {
+            return Verdict(
+                reports = Collections.unmodifiableList(reports),
```

Comment:
Nice, defending from mutations via public `Verdict.reports` getter. If it's created via constructor then it's not defended, so we could do that defense in the constructor or the getters. Or we can leave them undefended, planning to deprecate the constructors.

Reply:
I'd leave it as it is. Construction invocation is allowed within the library itself so we should be safe